spring-kafka/src/main/java/org/springframework/kafka/core/KafkaTemplate.java

```diff
@@ -307,13 +314,25 @@ protected void closeProducer(Producer<K, V> producer, boolean inLocalTx) {
 	 * @return a Future for the {@link RecordMetadata}.
 	 */
 	protected ListenableFuture<SendResult<K, V>> doSend(final ProducerRecord<K, V> producerRecord) {
+		return doSend(producerRecord, null);
+	}
+
+	/**
+	 * Send the producer record.
+	 * @param producerRecord the producer record.
+	 * @param callback the callback.
+	 * @return a Future for the {@link RecordMetadata}.
+	 */
+	protected ListenableFuture<SendResult<K, V>> doSend(final ProducerRecord<K, V> producerRecord,
+			final Callback callback) {
 		final Producer<K, V> producer = getTheProducer();
 		if (this.logger.isTraceEnabled()) {
 			this.logger.trace("Sending: " + producerRecord);
 		}
-		final SettableListenableFuture<SendResult<K, V>> future = new SettableListenableFuture<>();
+		final SettableListenableFuture<SendResult<K, V>> future = callback == null ? new SettableListenableFuture<>()
+				: null;
 		final boolean inLocalTx = inTransaction();
-		producer.send(producerRecord, new Callback() {
+		producer.send(producerRecord, callback != null ? callback : new Callback() {
```

Comment:
There is the block:
```
finally {
	if (KafkaTemplate.this.producers.get() == null) {
		closeProducer(producer, inLocalTx);
	}
}
```
In the end of `new Callback()` .
I think we just miss something there a bit and don't have choice unless wrap provided `callback` to that `new Callback()` any way. For consistency with other `send()` at least in case of TX...

WDYT?